### ISSUE TITLE
DistilBERT model on SST-2

### ISSUE THREAD

**Issue by sayakpaul** [sayakpaul/BERT-for-Mobile#2]
@khanhlvg

I was able to fine-tune the DistilBERT model on the SST-2 dataset. Note that if you train it on Colab GPU (P100) ([notebook](https://colab.research.google.com/github/sayakpaul/BERT-for-Mobile/blob/master/DistilBERT_SST-2_GPU.ipynb)) **_an epoch takes around half an hour_**. 

This is why I ported this model to TPU and executed the entire thing as a [Kaggle Kernel](https://www.kaggle.com/spsayakpaul/distilbert-sst-2). **_The model now takes around 100 seconds per epoch_**. Also, note that the TensorFlow version in the kernel is 2.2 because with 2.3 the TPU resolver does not work in the kernels. I have reported this issue to Martin Gorner. TFLite model files and the trained weights are available [here](https://github.com/sayakpaul/BERT-for-Mobile/releases/tag/v0.2.0). 

**Why not use Colab TPUs?**

This is because Colab TPUs require the dataset to be fetched from a GCS bucket. But in the case of Kaggle Kernels, this is not required. As long as the `GCS_DS_PATH` is set and your model is initialized within the TPU strategy scope you are good to go.  

As you'd notice that integer quantization fails for this one too. 

Let me know your thoughts. 

@farmaker47 @margaretmz FYI.

**Comment by sayakpaul** [comment]
@khanhlvg I evaluated the TensorFlow Lite models (DistilBERT) and the performance is very poor (**~50%**) i.e. to say the performance is completely random as it's a binary classification problem. Here's my [Colab Notebook](https://colab.research.google.com/github/sayakpaul/BERT-for-Mobile/blob/master/Evaluation_SST_2_DistilBERT.ipynb).

**Comment by khanhlvg** [comment]
Wow that's very surprising. What about the TF model before conversion?

**Comment by sayakpaul** [comment]
It's about 90% (validation accuracy). DistilBERT's (TFLite variant) performance is same (~50%) on the IMDB dataset as well :o

For MobileBERT it's `'accuracy': 0.9094036697247706` (the TFLite model).

**Comment by khanhlvg** [comment]
> It's about 90% (validation accuracy). DistilBERT's (TFLite variant) performance is same (~50%) on the IMDB dataset as well :o

Do you know if it's an issue with TF -> TFLite model conversion or there's a bug on pre-processing/post-processing in the TFLite evaluation pipeline? One way to find out is to compare directly the output of TF and TFLite model for the same input.

**Comment by sayakpaul** [comment]
> Do you know if it's an issue with TF -> TFLite model conversion or there's a bug on pre-processing/post-processing in the TFLite evaluation pipeline?

I don't think so. It's probably stemming from the conversion. 

> One way to find out is to compare directly the output of TF and TFLite model for the same input.

I checked it ([notebook](https://colab.research.google.com/gist/sayakpaul/8f3cf84ab6800f3bba0aade88a1bfb99/evaluation_sst-2_distilbert.ipynb)) and the results came out almost the same. `0.99993587` for the original model and `0.9999035` with the TFLite model.

**Comment by khanhlvg** [comment]
> I checked it (notebook) and the results came out almost the same. 0.99993587 for the original model and 0.9999035 with the TFLite model.

I'm confused. Do you mean that the TF and TFLite (without pre/post-processing) are the same, and there's no bug in the pre/post-processing pipeline. However the accuracy of the TF model is very different from the TFLite model?

**Comment by sayakpaul** [comment]
Sorry for the confusion. That's exactly what I meant.

**Comment by khanhlvg** [comment]
Hmm then it's weird because if each component (pre/post processing, model) works correctly, then where the discrepancy between TF and TFLite comes from...

Do you think we should pursue to find the root cause, or as the TFLite MobileBERT is performing quite well, we'll just go with it and ignore DistillBERT?

**Comment by sayakpaul** [comment]
> Hmm then it's weird because if each component (pre/post processing, model) works correctly, then where the discrepancy between TF and TFLite comes from...

Yes, exactly my thought. Could you review the following two notebooks once and see if I am right at all? 

* [Training and conversion ](https://github.com/sayakpaul/BERT-for-Mobile/blob/master/DistilBERT_SST-2_TPU.ipynb)
* [Evaluation](https://github.com/sayakpaul/BERT-for-Mobile/blob/master/Evaluation_SST_2_DistilBERT.ipynb)

> Do you think we should pursue to find the root cause, or as the TFLite MobileBERT is performing quite well, we'll just go with it and ignore DistillBERT?

I would say yes, we should since DistilBERT is a pretty popular model.

**Comment by khanhlvg** [comment]
> Yes, exactly my thought. Could you review the following two notebooks once and see if I am right at all?

I see. I'll take a look in a few weeks as it will require me understand DistillBERT well. I'm quite swamped with a few things at this moment.

**Comment by sayakpaul** [comment]
Cool. So, currently, I will flag it as blocked. I can also open an issue in the Hugging Face repository as well.

**Comment by khanhlvg** [comment]
> I can also open an issue in the Hugging Face repository as well.

Great idea! AFAIK they support TFLite natively so there's a good chance that they have faced this problem before.

**Comment by sayakpaul** [comment]
Done - https://github.com/huggingface/tflite-android-transformers/issues/10.

**Comment by sayakpaul** [comment]
@khanhlvg while debugging the code I found a mistake in my code. It's now corrected and the quantized (dynamic-range) DistilBERT model achieves about 90% validation accuracy. What is shocking to me is that the float16 quantized model resulted in ~49% accuracy. Here's the [notebook](https://github.com/sayakpaul/BERT-for-Mobile/blob/master/Evaluation_SST_2_DistilBERT.ipynb). 

I will report the benchmark results as soon as I can. Do you mind floating a document with an outline (if you'd like me to convert it to a blog post)?

**Comment by sayakpaul** [comment]
@khanhlvg I have filed an issue regarding the above - https://github.com/tensorflow/tensorflow/issues/43719. Currently, I am waiting for it to be answered. On a separate note, how would you structure the blog post if you'd like any? I think it might be very useful. 

Also, would you like to include the Question Answering task in it as well?
